src/random/mod.rs

```diff
@@ -20,15 +30,35 @@ impl RdRand {
         }
     }
 
-    /// Uniformly sampled u64
+    /// Uniformly sampled u64, u32, or u16
     #[inline]
-    #[cfg(target_arch = "x86_64")]
-    pub fn get(&self) -> u64 {
-        let res: u64;
+    pub fn get<T: RdRandPrimitive>(&self) -> Option<T> {
+        let res: T;
+        let ok: u8;
         unsafe {
-            asm!("rdrand %rax" : "={rax}"(res) ::: "volatile");
+            asm!("rdrand $0; setc $1;"
+                : "=r"(res) "=r"(ok)
+                :: "flags" : "volatile");
         }
-
-        res
+        match ok {
+            1 => Some(res),
+            _ => None
+        }
+    }
+    /// Uniformly sampled u64
+    #[cfg(target_arch = "x86_64")]
+    #[inline]
+    pub fn get_u64(&self) -> Option<u64> {
+        self.get()
+    }
+    /// Uniformly sampled u32
+    #[inline]
+    pub fn get_u32(&self) -> Option<u32> {
```

Comment:
This function should have `#[cfg(target_arch = "x86_64")]` too. It might work on x86, but fails on non-x86 platforms (this crate should be buildable on all platforms).

While `#[cfg(target_arch = "x86")]` would work too, I don't think that it's a good idea to add support for 32-bit targets because the Rest of the crate knows only two modes: x86_64 and non-x86_64.

Reply:
This makes sense. I've pulled x86_64 requirements up to the impl level, since everything in the impl would need it anyway.

Reply:
Thanks!